flink-core/src/main/java/org/apache/flink/core/fs/LimitedConnectionsFileSystem.java

```diff
@@ -0,0 +1,1097 @@
+			try {
+				originalStream.write(b, off, len);
+			}
+			catch (IOException e) {
+				handleIOException(e);
+			}
+		}
+
+		@Override
+		public long getPos() throws IOException {
+			try {
+				return originalStream.getPos();
+			}
+			catch (IOException e) {
+				handleIOException(e);
+				return -1; // silence the compiler
+			}
+		}
+
+		@Override
+		public void flush() throws IOException {
+			try {
+				originalStream.flush();
+			}
+			catch (IOException e) {
+				handleIOException(e);
+			}
+		}
+
+		@Override
+		public void sync() throws IOException {
+			try {
+				originalStream.sync();
+			}
+			catch (IOException e) {
+				handleIOException(e);
+			}
+		}
+
+		@Override
+		public void close() throws IOException {
+			if (closed.compareAndSet(false, true)) {
+				try {
+					originalStream.close();
+				}
+				catch (IOException e) {
+					handleIOException(e);
+				}
+				finally {
+					fs.unregisterOutputStream(this);
+				}
+			}
+		}
+
+		@Override
+		public void closeDueToTimeout() throws IOException {
+			this.timeoutException = new StreamTimeoutException();
+			close();
+		}
+
+		@Override
+		public boolean isClosed() {
+			return closed.get();
+		}
+
+		@Override
+		public long getLastCheckTimestampNanos() {
+			return lastCheckTimestampNanos;
+		}
+
+		@Override
+		public boolean checkNewBytesAndMark(long timestamp) throws IOException {
+			// remember the time when checked
+			lastCheckTimestampNanos = timestamp;
+
+			final long bytesNow = originalStream.getPos();
+			if (bytesNow > lastCheckBytes) {
+				lastCheckBytes = bytesNow;
+				return true;
+			}
+			else {
+				return false;
+			}
+		}
+
+		private void handleIOException(IOException exception) throws IOException {
+			if (timeoutException == null) {
+				throw exception;
+			} else {
+				// throw a new exception to capture this call's stack trace
+				throw new StreamTimeoutException(timeoutException);
+			}
+		}
+	}
+
+	/**
+	 * A data input stream that wraps a given data input stream and un-registers
+	 * from a given connection-limiting file system
+	 * (via {@link LimitedConnectionsFileSystem#unregisterInputStream(InStream)}
+	 * upon closing.
+	 */
+	private static final class InStream extends FSDataInputStream implements StreamWithTimeout {
+
+		/** The original data input stream to read from. */
+		private final FSDataInputStream originalStream;
+
+		/** The connection-limiting file system to un-register from. */
+		private final LimitedConnectionsFileSystem fs;
+
+		/** An exception with which the stream has been externally closed. */
+		private volatile StreamTimeoutException timeoutException;
+
+		/** The number of bytes written the last time that the {@link #checkNewBytesAndMark(long)}
+		 * method was called. It is important to initialize this with {@code -1} so that the
+		 * first check (0 bytes) always appears to have made progress. */
+		private volatile long lastCheckBytes = -1;
```

Comment:
`InStream` and `OutStream` have some overlap in the progress checking code. While they cannot have a common superclass for this functionality, logic and fields around `checkNewBytesAndMark(...)` and exception multiplexing could go to a separate class that is used in both streams to handle all the progress checks.

Reply:
Can try and do that...